runtime/jailer.go

```diff
@@ -0,0 +1,476 @@
+	srcFile, err := os.Open(src)
+	if err != nil {
+		return errors.Wrapf(err, "failed to open %v", src)
+	}
+	defer srcFile.Close()
+
+	if info, err := os.Stat(dst); err == nil {
+		// no error indicates the path already exists, we check to see if it is a
+		// directory we will concatenate the file name provided in source.
+		if info.IsDir() {
+			dst = filepath.Join(dst, filepath.Base(src))
+		}
+	} else if os.IsNotExist(err) {
+		// this occurs when a full path is provided including the file name
+	} else if err != nil {
+		return err
+	}
+
+	dstFile, err := os.OpenFile(dst, os.O_CREATE|os.O_WRONLY, mode)
+	if err != nil {
+		return errors.Wrapf(err, "failed to open %v", dstFile)
+	}
+	defer dstFile.Close()
+
+	_, err = io.Copy(dstFile, srcFile)
+	if err != nil {
+		return errors.Wrap(err, "failed to copy to destination")
+	}
+	return nil
+}
+
+func chownr(path string, uid, gid int) error {
+	return filepath.Walk(path, func(name string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		return os.Chown(name, uid, gid)
+	})
+}
+
+func (j runcJailer) jailerCommand(containerName string) *exec.Cmd {
+	cmd := exec.Command(j.Config.RuncBinaryPath, "run", containerName)
+	cmd.Dir = j.OCIBundlePath()
+	return cmd
+}
+
+// overwriteConfig will set the proper default values if a field had not been set.
+func (j runcJailer) overwriteConfig(cfg *Config, socketPath, configPath string) error {
+	spec := specs.Spec{}
+	configBytes, err := ioutil.ReadFile(configPath)
+	if err != nil {
+		return err
+	}
+
+	if err := json.Unmarshal(configBytes, &spec); err != nil {
+		return err
+	}
+
+	spec = j.setDefaultConfigValues(cfg, socketPath, spec)
+
+	spec.Root.Path = rootfsFolder
+	spec.Root.Readonly = false
+	spec.Process.User.UID = uint32(j.Config.UID)
+	spec.Process.User.GID = uint32(j.Config.GID)
+	configBytes, err = json.Marshal(&spec)
+	if err := ioutil.WriteFile(configPath, configBytes, 0666); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// setDefaultConfigValues will process the spec file provided and allow any
+// empty/zero values to be replaced with default values.
+func (j runcJailer) setDefaultConfigValues(cfg *Config, socketPath string, spec specs.Spec) specs.Spec {
+	if spec.Process != nil {
+		if spec.Process.Args == nil {
+			cmd := firecracker.VMCommandBuilder{}.
+				WithBin(cfg.FirecrackerBinaryPath).
+				WithSocketPath(socketPath).
+				Build(context.Background()) // TODO: use shim context here
+
+			spec.Process.Args = cmd.Args
+		}
+	}
+	return spec
+}
+
+// getUIDAndGID will the next free uid and gid that isn't being used by the
+// system.
+//
+// TODO: This can be removed once we jail with a user namespace
+func getUIDAndGID(parentCtx context.Context, config *Config) (uint32, uint32, error) {
+	ctx, cancel := context.WithTimeout(parentCtx, time.Second)
+	defer cancel()
+
+	// If idCount has not been set, we will set it to the minimum value
+	// TODO still a race need a channel or something to prevent
```

Comment:
I'm actually not sure we can really ever prevent race-conditions here without enormous effort. We would need to coordinate between different shim processes, so I'm not sure how to do that without some form of IPC or coordinated state management on the file system.

Reply:
Yea, I really had no idea either, but left in a comment to see if anyone had any ideas.